CONTRIBUTING.md

```diff
@@ -0,0 +1,181 @@
+
+#### .feature files
+
+The `.feature` files are written in Gherkin, a syntax cucumber uses. The "code" in there relies on the functions set up in `./lib/steps.js`.
+
+```js
+// The test_something.feature file step
+Then I sign
+
+// relies on the code in steps.js
+Then('I sign', { timeout: -1 }, async () => {
+  return wrapPromiseWithTimeout(
+    scope.steps.sign( scope ),
+    scope.timeout
+  );
+});
+```
+
+#### steps.js
+
+`./lib/steps.js` defines all the Gherkin sentences that our users (the developers) can write. See [cucumber documentation on defining steps](https://cucumber.io/docs/cucumber/step-definitions/?lang=javascript). We sometimes use "cucumber expressions" and, but often use regular expressions as we often want to be more flexible about the user's input.
+
+The file handles things like:
+* Creating state
+* Defining steps for the users/developers
+* Defining steps for internal testing, like testing for appropriate errors and error messages
+* Finishing a Scenario
+* Finishing all tests
+
+#### scope.js
+
+`./lib/steps.js` makes heavy use of `./lib/scope.js`, where most of the action happens. We have decided to mostly keep it in one file to avoid bouncing back and forth between files.
+
+Many `./lib/steps.js` steps make use of the `scope.steps` object. It is there to avoid duplicating code between steps.
+
+The file handles things like:
+* Finding form fields on a page
+* Filling out form fields
+* Generating random data for form fields
+* Waiting for pages to load
+* Checking for server restarts
+* Generating test report content
+
+### Logging
+
+`./lib/utils/log.js` takes care of logging information of various kinds (normal, debug, error) to the console in ways we find useful. We should probably find a logging library instead, but this is what we have for now. An important feature is that it ensures the messages clearly belong to ALKiln and not to some other process.
+
+### session_vars.js
+
+The `./lib/utils/session_var.js` file keeps track of what you might otherwise think of as environment variables. We like to think of them as constants, but some of them do need go through functions, so for consistency we get them all through functions. Also because it was easier to test them when they are functions.
+
+### Setup and takedown logic
+
+All of the setup and takedown logic is in the `./lib/docassemble` directory. It is named after the docassemble API code that is in there. Also, setup and takedown interact very closely with docassemble and the docassemble server.
+
+Other files also use the docassemble API functions, so this folder is in a bit of a messy position.
+
+### Bash commands
+
+Our `./package.json` uses the code in the `script` directive to setup, run, and takedown tests. We are working on converting this to a command line tool to avoid some complexity in GitHub actions, increase security, and align more with what this framework has grown into. (TODO: Define this better)
```

Comment:
```suggestion
Our `./package.json` uses the code in the `scripts` directive to setup, run, and takedown tests. We are working on converting this to a command line tool to avoid some complexity in GitHub actions, increase security, and align more with what this framework has grown into. (TODO: Define this better)
```